File: kmeans.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#define _GNU_SOURCE

double eucDist(double *pointA, double *pointB, int vecSize)
{
    double dist = 0;
    for (int i = 0; i < vecSize; i++)
    {
        dist = dist + ((pointA[i] - pointB[i]) * (pointA[i] - pointB[i]));
    }
    return sqrt(dist);
}

int main(int argc, char *argv[])
{
    /***** DATA FETCH *****/

    // arg size should be 5 for full, 4 for missing iteration
    char *err = NULL;
    double EPSILON = 0.001;
    int iter = 100;
    int k = strtol(argv[1], err, 10); // K should be int
    int pointIndex = 0;
    if (argc == 6 || argc == 7)
    {
        if (argc == 6)
        {
            pointIndex = 5;
        }
        else
        {
            pointIndex = 6;
        }
    }
    else
    {
        printf("An Error Has Occurred");
        return 1;
    }

    if (err == argv[1] || *err != '\0')
    {
        printf("Invalid number of clusters!");
        return 1;
    }
    else
    {
        // k is integer
        if (k <= 0)
        {
            printf("Invalid number of clusters!");
            return 1;
        }
    }
    int n = strtol(argv[2], err, 10); // N should be int
    if (err == argv[2] || *err != '\0')
    {
        printf("Invalid number of points!");
        return 1;
    }
    else
    {
        // k is integer
        if (n <= 0)
        {
            printf("Invalid number of points!");
            return 1;
        }
        if (k >= n)
        {
            printf("Invalid number of clusters!");
            return 1;
        }
    }
    int d = strtol(argv[3], err, 10); // d should be int
    if (err == argv[3] || *err != '\0')
    {
        printf("Invalid dimension of point!");
        return 1;
    }
    else
    {
        // d is integer
        if (d <= 0)
        {
            printf("Invalid dimension of point!");
            return 1;
        }
    }
    if (argc == 5)
    {
        iter = strtol(argv[4], err, 10); // if iter present it should be int
        if (err == argv[4] || *err != '\0')
        {
            printf("Invalid maximum iteration!");
            return 1;
        }
        else
        {
            if (iter >= 1000 || iter <= 0)
            {
                printf("Invalid maximum iteration!");
                return 1;
            }
        }
    }
    /***** FETCH POINTS *****/
    double **points = (double **)malloc(n * sizeof(double *)); // the actual points
    int *direct = malloc(n * sizeof(int));                     // points to in what cluster each point belongs
    int *clusterSize = malloc(k * sizeof(int));                // saves cluster size
    for (int i = 0; i < n; i++)
    {
        direct[i] = -1;
    }
    for (int i = 0; i < n; i++)
    {
        points[i] = (double *)malloc(d * sizeof(double));
    }
    int read;
    char *line = NULL;
    size_t len = 0;
    int g = 0;
    const char *delim = " ,\n";
    for (int i = 0; i < n; i++)
    {
        // read the entire line
        read = getline(&line, &len, stdin);
        if (read == -1)
        {
            free(line);
            free(direct);
            free(clusterSize);
            for (int i = 0; i < n; i++)
            {
                free(points[i]);
            }
            free(points);
            printf("An Error Has Occurred");
            return 1;
        }
        g = 0;
        char *token = strtok(line, delim);
        while (token != NULL)
        {
            points[i][g] = strtod(token, NULL);
            token = strtok(NULL, delim);
            g++;
        }
    }
    free(line);
    /***** INIT CENTER *****/
    double **centers = (double **)malloc(k * sizeof(double *));
    for (int i = 0; i < k; i++)
    {
        centers[i] = (double *)malloc(d * sizeof(double));
    }

    double **prevCenters = (double **)malloc(k * sizeof(double *));
    for (int i = 0; i < k; i++)
    {
        prevCenters[i] = (double *)malloc(d * sizeof(double));
    }
    for (int i = 0; i < k; i++)
    {
        for (int j = 0; j < d; j++)
        {
            centers[i][j] = points[i][j];
        }
    }
    /***** ITERATION LOOP *****/
    int flag = 1;
    int bestIndex = 0;
    double bestDist = 0.0;
    double currentDist = 0.0;
    while (flag != 0 && iter > 0)
    {
        // prepare prev centers
        for (int i = 0; i < k; i++)
        {
            for (int j = 0; j < d; j++)
            {
                prevCenters[i][j] = centers[i][j];
            }
        }
        for (int i = 0; i < n; i++)
        {
            clusterSize[i] = 0;
        }
        flag = 0;
        // direct points to clusters
        for (int i = 0; i < n; i++)
        {
            bestDist = -1;
            bestIndex = 0;
            for (int j = 0; j < k; j++)
            {
                currentDist = eucDist(points[i], centers[j], d);
                if (bestDist == -1 || bestDist > currentDist)
                {
                    bestDist = currentDist;
                    bestIndex = j;
                }
            }
            direct[i] = bestIndex;
            clusterSize[bestIndex] = clusterSize[bestIndex] + 1;
        }
        // make new centers
        // step 1 clean centers
        for (int i = 0; i < k; i++)
        {
            for (int j = 0; j < d; j++)
            {
                centers[i][j] = 0.0;
            }
        }
        // step 2 adjust new center raw values
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < d; j++)
            {
                centers[direct[i]][j] = centers[direct[i]][j] + points[i][j]; // direct directs each point to a relevant center so direct[i] will give point[i] its cluster index
            }
        }
        // step3 divide each coord by the size
        for (int i = 0; i < k; i++)
        {
            for (int j = 0; j < d; j++)
            {
                centers[i][j] = centers[i][j] / ((double)clusterSize[i]);
            }
        }
        // test new centers against prev centers
        for (int i = 0; i < k; i++)
        {
            if (eucDist(centers[i], prevCenters[i], d) >= EPSILON)
            {
                flag = 1;
            }
        }
    }
    /***** PRINT CENTERS *****/
    for (int i = 0; i < k; i++)
    {
        for (int j = 0; j < d; j++)
        {
            printf("%.4f", centers[i][j]);
            if (j < d - 1)
            {
                printf(",");
            }
        }
        printf("\n");
    }
    /***** FREE MEMORY *****/
    free(direct);
    free(clusterSize);
    for (int i = 0; i < n; i++)
    {
        free(points[i]);
    }
    free(points);
    for (int i = 0; i < k; i++)
    {
        free(centers[i]);
        free(prevCenters[i]);
    }
    free(centers);
    free(prevCenters);
    return 0;
}

File: kmeans.py
import copy
import math
k = 3
n=800
d=3
iter = 600
path:str = "input_1.txt"
nodes = [] # will be 2D list
EPSILON = 0.001

def handleErrors(msg:str ="An Error Has Occurred"):
    print(msg)
    SystemExit(1)

def verifyData():
    if n<1 or not isinstance(n,int): handleErrors("Invalid number of points!")
    if 1 > k or k>n: handleErrors("Invalid number of clusters!")
    if not isinstance(d,int) or d<1: handleErrors("Invalid dimension of point!")
    if iter <= 1 or iter >=1000: handleErrors("Invalid maximum iteration!")


def euclidianDist(pointA:list,pointB: list)->float:
    if(len(pointA) != len(pointB)): handleErrors()
    dist = 0
    for i in range(0,len(pointA)):
        dist =math.pow((pointA[i]-pointB[i]),2) + dist
    return math.sqrt(dist)

def readFromFile(path: str):
    try:
        f = open(path, "r")
        for line in f:
            coords = line.split(',')
            newNode = []
            try:
                for coord in coords:
                    newNode.append(float(coord))
            except:
                handleErrors()
            nodes.append(newNode)
    except:
        handleErrors()
# Start of Script
verifyData()
# Read file
readFromFile(path)
# Initilize Centers for the first time
clusters = [[] for i in range(0,k)]

centers = [nodes[i] for i in range(0,k)]
prevCenters = copy.deepcopy(centers)

for i in range(0,k):
    clusters[i].append(centers[i])

def generateCenters():
    for index,cluster in enumerate(clusters):
        newCenter = [0 for i in range(0,d)]
        for i in range(0,d): # coord index
            for j in range(0,len(cluster)): # point index
                newCenter[i] = (newCenter[i] + cluster[j][i])
            newCenter[i] = newCenter[i]/len(cluster)
        centers[index] = newCenter

def calcClusters():
    for point in nodes:
        index = 0
        dist:float = -1
        for i in range(0,len(centers)):
            if dist == -1:
                dist = euclidianDist(centers[i],point)
            else:
                newDist =euclidianDist(centers[i],point)
                if newDist  < dist:
                    dist = newDist
                    index = i
        clusters[index].append(point)
# Iteration loop 
while iter>0:
    iter = iter-1
    calcClusters()
    generateCenters()
    for i in range(0,len(centers)):
        if euclidianDist(centers[i],prevCenters[i]) > EPSILON:
            prevCenters = copy.deepcopy(centers)
            break
    else:
        break
print(iter)
print(centers)
#format output